### ISSUE TITLE
[Enhancement, Placeholder] Bottom Tab Bar on Android or Top Tab Bar on iOS

### ISSUE THREAD

**Issue by kingces95** [xamarin/Xamarin.Forms#1675]
## Rationale

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.

## Implementation

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.
```csharp
```

## Expected Result

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.

## Implications for CSS

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.

## Backward Compatibility

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.

## Difficulty : Easy

Lorem ipsum accumsan etiam congue sed nostra congue, a in aliquet orci habitasse maecenas est suscipit, diam auctor arcu posuere massa augue.

**Comment by davidortinau** [comment]
Related #1400

**Comment by mikescandy** [comment]
Hi, I just opened a pull request on this issue. I have a few questions about the implementation, but it is a starting point.
The mapping was obvious enough.

**Comment by lemieuxs** [comment]
I am reallly looking forward this pull-request being merge into Xamarin Forms. Is there an E.T.A or something. I haven't found any viable solution to have tabbed page at the bottom on Android.

**Comment by rcunniffe** [comment]
@mikescandy any update on this?

**Comment by b1mn** [comment]
When this will be release to try?

**Comment by topfunet** [comment]
Finally, I am looking farword to this so long!
Thank you so much!

**Comment by niels9001** [comment]
When will this enhancement roll out?

**Comment by rcunniffe** [comment]
+1 
can I get it on nightly build? cc @PureWeen @davidortinau

**Comment by niels9001** [comment]
And how does it work? Can we set it in the XAML of our Forms app as well (as a property of the TabbedPage)?

**Comment by rcunniffe** [comment]
@netonjm I think it's just a prop "UseBottomNavigation" https://github.com/mikescandy/Xamarin.Forms/blob/27c6da20b0ade7f96c0fd889f00c96b7e4db1fcc/Xamarin.Forms.Platform.Android/AppCompat/TabbedPageRenderer.cs#L60

but not really sure of the details (icons and colours).

**Comment by netonjm** [comment]
It's me :-)

**Comment by rcunniffe** [comment]
oops

**Comment by PureWeen** [comment]
@niels9001   you can see all the options for setting up the icons etc here

https://github.com/xamarin/Xamarin.Forms/blob/67d2be4cd224678a350f258b888c568ac8898a8f/Xamarin.Forms.Core/PlatformConfiguration/AndroidSpecific/TabbedPage.cs

**Comment by niels9001** [comment]
Awesome! So by setting "UseBottomNavigation" in XAML it will show the tabbar at the bottom on Android and iOS (like it already does for a TabbedPage), right?

Awesome :D! Is this enhancement already part of the latest pre-release NuGet package? And if not, any ETA on this?

**Comment by rcunniffe** [comment]
@niels9001 and @PureWeen I've installed xamarin forms `3.1.0.506097` upgrading for version 2. Seems to be working now but am i right in thinking this isn't available there yet?

Any chance anybody could explain the process of when cards (on the done closed column etc) get merged into the nightly from here https://github.com/xamarin/Xamarin.Forms/projects/2#card-8921124? 

Also how does this this board relate? https://github.com/xamarin/Xamarin.Forms/projects/6 

thanks!

**Comment by rcunniffe** [comment]
ok went through the pull requests and ended up doing this to get it to work on xamarin.forms

```
private Xamarin.Forms.TabbedPage __bottomBarPage;
        private INavigationService __navigationService;
        private Page __currentPage;

        public BottomTabBarPageBuilder(INavigationService navigationService)
        {
            __navigationService = navigationService;
            __bottomBarPage = new Xamarin.Forms.TabbedPage();
            __bottomBarPage.BarTextColor = (Color)App.Current.Resources["tile2"]; // Setting Color of selected Text and Icon
            __bottomBarPage.On<Xamarin.Forms.PlatformConfiguration.Android>().SetToolbarPlacement(ToolbarPlacement.Bottom);
            // You can only define the color for the active icon if you set the Bottombar to fixed mode
            //__bottomBarPage.Navigation..FixedMode = true;
        }|
```

nevermind the my code but the main part is `            __bottomBarPage.On<Xamarin.Forms.PlatformConfiguration.Android>().SetToolbarPlacement(ToolbarPlacement.Bottom);` is this the right way?

**Comment by PureWeen** [comment]
Yes that's the right way. It's a platform specific thing so just need to set that before the control is rendered.

**Comment by maherzaidoune** [comment]
hey, does this support fixed mode ?!  i can't find a way to set it

**Comment by PureWeen** [comment]
@maherzaidoune  not yet. 
I had looked into a little bit for initial implementation but decided to leave it for the next version.

Just to be clear you're talking about this right?
https://stackoverflow.com/questions/41649494/how-to-remove-icon-animation-for-bottom-navigation-view-in-android

**Comment by niels9001** [comment]
I'm a bit puzzled. Where are we supposed to set the ToolBarPlacement?

`namespace Test.Mobile.Views
{
    [XamlCompilation(XamlCompilationOptions.Compile)]
    public partial class TestPage : Xamarin.Forms.TabbedPage
    {
       
        public Test()
        {
            InitializeComponent();
       
            On<Xamarin.Forms.PlatformConfiguration.Android>().SetToolbarPlacement(ToolbarPlacement.Bottom);

        }`

Like this?

**Comment by b1mn** [comment]
How does `On<Android>().SetElevation();` work? I tried to set it with some float value does not see any effect. I wanted to drop some shadow to distinguish between page content and the bottom bar.

**Comment by PureWeen** [comment]
@niels9001 yes

Is that working for you ok?

**Comment by niels9001** [comment]
Nope, I'm getting this error 

> Unhandled Exception:

System.TypeLoadException: Could not resolve type with token 0100008e (from typeref, class/assembly Xamarin.Forms.PlatformConfiguration.AndroidSpecific.ToolbarPlacement, Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null) occurred

**Comment by 15mgm15** [comment]
I am amazed. Brilliant. @PureWeen Thanks so much for this.

**Comment by 15mgm15** [comment]
@PureWeen  Please let me know if I can help with the Badge implementation.

**Comment by niels9001** [comment]
@PureWeen Thanks, that worked :)!

How do we set the background color of the entire bar? Or enable a dropshadow?

**Comment by bdgza** [comment]
If the child page (tab) of the bottom TabbedPage is a NavigationPage then it, and its contents (ContentPage with ScrollView), is rendered underneath the tab bar.

**Comment by DevEddy** [comment]
Nice work!
But why 5 children is the max? It crashes when I try to add more.

**Comment by PureWeen** [comment]
Yea, I went back and forth with the best way to help the user with that. If you look there's a GetMaxCount you can call which right now just returns a 5 for the bottom tabs and a max int for top tabs. I'll look to see if I can propagate a more useful error message :-)

**Comment by PureWeen** [comment]
@bdgza 

Thank you for reporting. I've created an issue for this
https://github.com/xamarin/Xamarin.Forms/issues/2993

**Comment by niels9001** [comment]
@PureWeen BarBackgroundColor doesn't seem to work to change the background color of the bar when placed at the Bottom on Android?

**Comment by PureWeen** [comment]
@niels9001  do you have a reproduction for that by chance?

I tested and for me it works

![image](https://user-images.githubusercontent.com/5375137/41320655-d833cb18-6e5d-11e8-9465-33a56cd7d7cb.png)

**Comment by wireless-r** [comment]
How to prevent the icon animation? The icon sliding/resizing, and the text visibility changing.

**Comment by PureWeen** [comment]
@dstarec  a new nightly for 3.2 is up on myget

Currently there is not but what you're asking is for this
https://github.com/xamarin/Xamarin.Forms/issues/2948
right?

**Comment by LynoDesu** [comment]
Sorry if I've missed this somewhere, but is there a way to disable the "Icon Zooming" effect, for the selected tab? I can see if you screenshot above, the selected tab icon isn't "zoomed"

**Comment by PureWeen** [comment]
@LynoDesu   @amrkamal1993   will address in an upcoming release
https://github.com/xamarin/Xamarin.Forms/issues/3083

**Comment by LynoDesu** [comment]
@amrkamal1993 I have gotten around it by using a custom renderer. I'll post my code later when I'm home

**Comment by LynoDesu** [comment]
@amrkamal1993 @maherzaidoune 
See gist here:
https://gist.github.com/LynoDesu/b29aaec1c119baaabfdba4f4a89bd3f4

**Comment by LynoDesu** [comment]
How have you implemented it? With my code, you need to use the custom control BottomNavTabPage instead of the TabbedPage:

`<?xml version="1.0" encoding="utf-8" ?>
<controls:BottomNavTabPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:views="clr-namespace:MyProject.App.Views;assembly=MyProjects.App"
             xmlns:controls="clr-namespace:MyProject.App.Controls;assembly=MyProjects.App"
             x:Class="MyProject.App.Views.MainTabPage"
             Title="">
    <views:NewsFeed></views:NewsFeed>
    <views:Rewards></views:Rewards>
    <views:Nominations></views:Nominations>
    <views:Notifications></views:Notifications>
</controls:BottomNavTabPage>`

![disabledshiftmode](https://user-images.githubusercontent.com/17926055/41710716-42c52486-753e-11e8-8185-3c2b6d1e6e4b.png)

**Comment by nhdanh** [comment]
How can I add Badge Text in bottom tab ?

**Comment by amrkamal1993** [comment]
@LynoDesu  thx it works finally !! ^_^

**Comment by amrkamal1993** [comment]
@nhdanh  https://github.com/xabre/xamarin-forms-tab-badge
this may help you :)

**Comment by nhdanh** [comment]
@amrkamal1993 : thank you, but I try this plugin. It not add badge text when I use Bottom bar.

**Comment by amrkamal1993** [comment]
@nhdanh  tab is still the same as top one it should work with the new feature .. bottom tab placement

**Comment by nhdanh** [comment]
@amrkamal1993  when I set On<Xamarin.Forms.PlatformConfiguration.Android>().SetToolbarPlacement(ToolbarPlacement.Bottom). It not render Badge text. I remove this setting bottom. It will display Badge text

**Comment by nhdanh** [comment]
anything else about badge text ?

**Comment by syed-pelican** [comment]
@LynoDesu Awesome! Your solution just helped made my day! Thanks!

**Comment by aalbendin** [comment]
I tried to update Xamarin.Forms to the latest version and I get this error when I try to set the ToolbarPlacement in my TabbedPage class:

> Unhandled Exception:
> 
> System.TypeLoadException: Could not resolve type with token 01000029 (from typeref, class/assembly Xamarin.Forms.PlatformConfiguration.AndroidSpecific.ToolbarPlacement, Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null) ocurrió

I dont know what im doing wrong, I think this error is related to the versions of the nuget packages. What can I do to resolve it?

**Comment by PureWeen** [comment]
@aalbendin  that will happen if something is out of date or just quite isn't building the correct libraries. I would recommend deleting all your bin/obj folders and then rebuilding your project and that should fix it.

A user further up was having the same issue and that resolved it for them

https://github.com/xamarin/Xamarin.Forms/issues/1675#issuecomment-395211876

**Comment by cassionandi** [comment]
is this available on stable 3.1.0.583944 version? I have added the On<Xamarin... code on my TabbedPage but nothing happened.

**Comment by 15mgm15** [comment]
Anyone else trying to hide the TabbedNavigationBar? I am having an issue on Android if I set the `VisibiltyStates.Gone` to the `BottomNavigationView ` the view is not completely gone:

![screenshot_1530818022](https://user-images.githubusercontent.com/3074404/42343122-aaed2764-805d-11e8-9374-eef52f0093f7.png)

**Comment by PureWeen** [comment]
@15mgm15   that issue is semi  related to
https://github.com/xamarin/Xamarin.Forms/issues/3055

The ViewPager that displays the content accounts for the height of the bottom navigation bar so that content doesn't get hidden behind it.  In order for that content to be taken up again it has to re-layout the viewpager but there's nothing right now that watches for that visibility and change 

What's your use case?  Could you instead Push a Modal page onto the stack so that it just overlays the entire screen instead of trying to hide the bottom nav bar?

**Comment by 15mgm15** [comment]
@PureWeen Thanks for the response! unfortunately using modal pages breaks the client design, so I will have to find a solution.

**Comment by PureWeen** [comment]
@15mgm15  if you use a top nav bar does it do the same thing? Or does this only happen with the bottom nav bar?

**Comment by 15mgm15** [comment]
@PureWeen Actually let me check and let you know.

**Comment by 15mgm15** [comment]
@PureWeen Using the normal tav bar everything works as expected, the `TabLayout` of the `TabbedPageRenderer` is hidden or showed as we pleased. But I tried setting the height to 0 on the layout parameters and it actually worked! 

`BottomNavigationView _bottomBar;
...
if (_extendedTabbedPage.BottomTabBarHidden)
{
    _layoutParams.Height = 0;
    _bottomBar.LayoutParameters = _layoutParams;
}
else
{
    _layoutParams.Height = _bottomBarHeight;
    _bottomBar.LayoutParameters = _layoutParams;
}`

**Comment by PureWeen** [comment]
Hey that's good to hear!! Once https://github.com/xamarin/Xamarin.Forms/issues/3055 gets fixed I'm hoping that will smooth out some of the rough parts a bit

@cassionandi  are you still having issues? If you set the navbar location in your constructor or use XAML you shouldn't be having any issues

**Comment by niels9001** [comment]
If you do not set a label, and only the icon, the icons are not center aligned vertically. It looks weird, like there is a label missing.

Is this expected behaviour? Would be nice if the icons would be in the center (like the Outlook app on Android).

**Comment by bverp** [comment]
I think I'm missing something. I updated Xamarin Forms to release 3.1.0.583944 but I get two errors on the line  `On<Xamarin.Forms.PlatformConfiguration.Android>().SetToolbarPlacement(ToolbarPlacement.Bottom);`

**The name ToolbarPlacement' does not exist in the current context**

 and 

**Error CS1061: 'IPlatformElementConfiguration<Android, TabbedPage>' does not contain a definition for 'SetToolbarPlacement' and no extension method 'SetToolbarPlacement' accepting a first argument of type 'IPlatformElementConfiguration<Android, TabbedPage>' could be found** 

any suggestions?

**Comment by PureWeen** [comment]
@bverp 

```
using Xamarin.Forms.PlatformConfiguration.AndroidSpecific
```
?

**Comment by 15mgm15** [comment]
If anyone wants to **override the OnTabReselected** event or wants to **add badges** to the tabs or wants to **hide the BottomNavigationView** here is an example:

https://gist.github.com/15mgm15/76914bf623e33676252864ac1898e7f9

**Comment by niels9001** [comment]
@bverp Close VS, delete the obj and bin folder and restart the project. That should work, atleast it did for me.

**Comment by taramasalata** [comment]
@niels9001 that didn't work for me. @bverp did you get it working? I tried downloading and building this project, but get the error: The name ToolbarPlacement' does not exist in the current context
https://github.com/15mgm15/XamarinFormsBottomTabbedPage

**Comment by taramasalata** [comment]
I'm doing my nut a bit - does anyone have time to help?
I tried a number of things to get bottom tabs in Android, including both options in this link. Whatever I do, my solution does not recognise the update.
https://developer.xamarin.com/releases/xamarin-forms/xamarin-forms-3.1/3.1.0/
I have forms version 3.1.0.637273 (and have tried the last 5 releases/prereleases too).
I have cleaned, built, rebuilt, deleted the bin/obj directories - I cannot seem to see this update at all no matter what I do. 

For example, this is my current XAML

<TabbedPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:ProjName"
             x:Class="ProjName.MainPage"
    xmlns:android="clr-namespace:Xamarin.Forms.PlatformConfiguration.AndroidSpecific;assembly=Xamarin.Forms.Core"
            android:TabbedPage.ToolbarPlacement="Bottom"
            android:TabbedPage.BarItemColor="Black"
            android:TabbedPage.BarSelectedItemColor="Red">
    
    <local:LoginPage Title="Favorites" Icon="ic_favorites.png" />
    <local:LoginPage Title="Friends" Icon="ic_friends.png" />
    <local:LoginPage Title="Nearby" Icon="ic_nearby.png" />
    <local:LoginPage Title="Recents" Icon="ic_recents.png" />
    <local:RegistrationPage Title="Restaurants" Icon="ic_restaurants.png" />
</TabbedPage>

and the result is

No property, bindable property, or event found for 'ToolbarPlacement', or mismatching type between value and property.		

Similar problem doing it programatically - it simply does not recognise that update.

I'm on Windows VS, using 8.1 Android for testing. If anyone can tell me what I'm missing it would make me very happy.

@pureween?

**Comment by davidortinau** [comment]
@taramasalata confirm that all your projects are using the most recent version of 3.1.0. Something is not matching up.

You can also check my sample project here: https://github.com/davidortinau/TheLittleThingsPlayground

I also confirmed I could do the bottom tabs from a shared project without error. Project: 
[App25.zip](https://github.com/xamarin/Xamarin.Forms/files/2194141/App25.zip)

**Comment by taramasalata** [comment]
@davidortinau Thank you for helping. I have tried variations of this in 8 different solutions.  I have updated all projects to 3.1.0 and triple checked everything. I have checked the "updates" tab for each one repeatedly they are all up to date. I can't think what else to try. I'll try your sample projects in the morning and get back to you. Maybe I should try reinstalling VS.

**Comment by taramasalata** [comment]
Your app25 will not build with error : No property, bindable property, or event found for 'ToolbarPlacement', or mismatching type between value and property.			

Are you on Windows or Mac? 

I'm going to try a complete reinstall of VS now I really cannot think what else to try

**Comment by taramasalata** [comment]
Nope. No joy with fresh install, running App25. This is so frustrating.
 error: Position 6:13. No property, bindable property, or event found for 'ToolbarPlacement', or mismatching type between value and property.

**Comment by taramasalata** [comment]
Are there any packages I need for this beside Xamarin.Forms?

**Comment by PureWeen** [comment]
Hmm something in your build tool chain just isn't quite right @taramasalata 
Maybe set the build output to verbose to see any messages?  Check your Android Sdk make sure you have latest build tools?  Install 4.7.1 sdk see if that resolves? delete all your cached Xamarin.forms nugets 
c:\users\<name>\.nuget

Use something like ILSpy or dotpeek to look at the dlls in your debug directory and check the version or see if you can find those values in there?

Try a code base approach instead of XAML?

**Comment by taramasalata** [comment]
Thank you @pureween. I tried both code based and xaml. I also tried reinstalling VS and tried the "preview" version too. It's likely my uninstall/reinstall didn't delete the .nuget files though. I'll try all your suggestions now.

**Comment by taramasalata** [comment]
Having reinstalled all things Android this now works. Thanks so much all!

Quick Q am I correct in thinking this will only run on Android 8.1 up?

**Comment by ricardouffcomp** [comment]
Hi @taramasalata. I had a similar issue and it got fixed using the following steps:
- Deleting all "bin" and "obj" folders of my app
- Click on "Clean Solution"
- Restart Visual Studio 
- Click on "Rebuild Solution".

**Comment by PureWeen** [comment]
@taramasalata   YAY!!!!  It should run on any device with at least API 15

In building this I tested against API 15, 19, 23, 27

Getting the tint colors working on 19 and 15 was really fun

**Comment by amrkamal1993** [comment]
Can I Use Bottom Tab with Master Detail ?

**Comment by bverp** [comment]
@taramasalata yes i got it working after adding `using Xamarin.Forms.PlatformConfiguration.AndroidSpecific` and restarting VS and cleaning the solution.

**Comment by marcelinhovt** [comment]
When use on Top, works with 6 Childrens. When is at the Bottom, just 5.
The app crashs with the message "Java.Lang.IllegalArgumentException: <Timeout exceeded getting exception details>"

**Comment by PureWeen** [comment]
@marcelinhovt  yea that's an android limitation.

The BottomNavBar in Android only supports 5 items and there really isn't anyway for us to get around this

**Comment by marcelinhovt** [comment]
@PureWeen  oh...i didn´t know. Thanks... I will try to write some render to extend

**Comment by PureWeen** [comment]
@marcelinhovt  so what I mean is that natively the BottomNavigationBar in android doesn't support more then 5 items.  That's not a limitation imposed by our design

**Comment by marciosouzajunior** [comment]
There is a way to set elevation on bottom tab bar?

**Comment by amrkamal1993** [comment]
I Have a new problem in android limitation even i have only 5 tabs app is crashing

**Comment by PureWeen** [comment]
@amrkamal1993  if you could create a repro and a new issue then we can hopefully address the issue

**Comment by PureWeen** [comment]
@amrkamal1993 all the test code we have for bottom tab uses icons.

Can you post a repro or code I can test to see if I can recreate the crash?

**Comment by pakbaz** [comment]
Is there a way to achieve equal distance between icons in the bottom similar to what iOS does? right now, the behavior is to zoom whatever icon tab you are on and display text for it and having other icons positioned differently. It looks weird in some setup specially if you have 4 or 5 tabs.
